### ISSUE TITLE
Batch menu for Backup and Restore in 1-click ops

### ISSUE THREAD

**Issue by Cyberavater** [MuntashirAkon/AppManager#136]
Update: More detailedly described here; https://github.com/MuntashirAkon/AppManager/issues/136#issuecomment-719652773

2 batch menu with specific conditions like other backup managers have should be very useful.
Backup and Restore Drawer in 1-click ops should be useful.

![Screenshot_20201030-212601583 (1)](https://user-images.githubusercontent.com/32187927/97724685-6e9d9f00-1af7-11eb-8351-df6d27e7bb2c.jpg)

**Comment by MuntashirAkon** [comment]
Please be more specific and clear about exactly what feature you want.

**Comment by Cyberavater** [comment]
Hmm, I'll contact you in TG then. The SS should be sufficient to understand, Thanks!

**Comment by MuntashirAkon** [comment]
> The SS should be sufficient to understand, Thanks!

Sorry, it wasn't. Feature requests usually consist of a description of features, current alternatives and additional context. You've only provided additional context.

**Comment by MuntashirAkon** [comment]
_[Continued from TG]_

So what you want is two 1-click ops:
1. **Backup apps:** Options regarding backing up apps
  a. Backup all apps
  b. Backup apps without any previous backups
  c. Backup updated apps
  d. Backup apps with existing backups by detecting changes (not sure why this is needed)
2. **Restore apps:** Options regarding restoring already backed up apps
  a. Restore all apps
  b. Restore apps that are not currently installed
  c. Restore apps whose backup version is higher than the installed version

**Comment by Cyberavater** [comment]
> _[Continued from TG]_
> 
> So what you want is two 1-click ops:
> 
> 1. **Backup apps:** Options regarding backing up apps
>    a. Backup all apps
>    b. Backup apps without any previous backups
>    c. Backup updated apps
>    d. Backup apps with existing backups by detecting changes (not sure why this is needed)
> 2. **Restore apps:** Options regarding restoring already backed up apps
>    a. Restore all apps
>    b. Restore apps that are not currently installed
>    c. Restore apps whose backup version is higher than the installed version

Yup! that's it! If you still have some questions feel free to ask away,

**Comment by MuntashirAkon** [comment]
> If you still have some questions feel free to ask away,

Yes. About option 1(d). Why we need this option? As I understand, it means that new backups have to be made for each apps, the matched them with the existing backups (via checksums) to see if there's any difference, and if there _is_ any difference, the existing backups have to be replaced with the new one. This seems to be an overhead since backups have to made in one way or the other.

**Comment by Cyberavater** [comment]
> _[Continued from TG]_
> 
> So what you want is two 1-click ops:
> 
> 1. **Backup apps:** Options regarding backing up apps
>    a. Backup all apps
>    b. Backup apps without any previous backups
>    c. Backup updated apps
>    d. Backup apps with existing backups by detecting changes (not sure why this is needed)
> 2. **Restore apps:** Options regarding restoring already backed up apps
>    a. Restore all apps
>    b. Restore apps that are not currently installed
>    c. Restore apps whose backup version is higher than the installed version

While you work on it, I'd like to add that even if you select one option from those, there should be another page after that'll only contain the apps which fulfill the condition (other apps won't be in this list). All apps on that list should be selected, now if the user wants to exclude one app from the process he'll unselect it and proceed with the op.

**Comment by Cyberavater** [comment]
> > If you still have some questions feel free to ask away,
> 
> Yes. About option 1(d). Why do we need this option? As I understand, it means that new backups have to be made for each apps, then matched them with the existing backups (via checksums) to see if there's any difference, and if there _is_ any difference, the existing backups have to be replaced with the new one. This seems to be an overhead since backups have to made in one way or the other.

No, it shouldn't make a backup first and then check with the previous one. TB and SB have this feature and they do the check _before_ they make the backup.

**Comment by Cyberavater** [comment]
> > _[Continued from TG]_
> > So what you want is two 1-click ops:
> > 
> > 1. **Backup apps:** Options regarding backing up apps
> >    a. Backup all apps
> >    b. Backup apps without any previous backups
> >    c. Backup updated apps
> >    d. Backup apps with existing backups by detecting changes (not sure why this is needed)
> > 2. **Restore apps:** Options regarding restoring already backed up apps
> >    a. Restore all apps
> >    b. Restore apps that are not currently installed
> >    c. Restore apps whose backup version is higher than the installed version
> 
> While you work on it, I'd like to add that even if you select one option from those, there should be another page after that'll only contain the apps which fulfill the condition (other apps won't be in this list). All apps on that list should be selected, now if the user wants to exclude one app from the process he'll unselect it and proceed with the op.

For option 1(d) all apps are on the list.

**Comment by MuntashirAkon** [comment]
> No, it shouldn't make a backup first and then check with the previous one.

Then, what changes will be detected here?

> TB and SB have this feature and they do the check _before_ they make the backup.

What do they check?

**Comment by MuntashirAkon** [comment]
> For Apk it's signature and version. If both matches then it skips backing up the apk and moves to data.

Version checks should be covered by 1(c). As for signature check, things are a bit complicated when encryption is enabled (which might be enforced in the final release). The only unencrypted file is the metadata file which doesn't contain any checksums. Checksums are stored in the **checksums.txt** file (which also contain the checksum of the metadata to make sure that the metadata file isn't tempered with). So, to check signatures, **checksums.txt** has to be decrypted first (decryption takes time depending on the algorithm and device used).

> And for data, it seems they check if the app was launched after the previous backup was made,

This could be done.

> if yes then it backs up data (or maybe size?), then for obb I can't even guess, as I haven't noticed while backing up such app, but I guess checksums should be good enough since it's only one or two files (or there is some faster check? IDK pinpoint size?). And for external data, only size check seems fast and reliable choice considering there are thousands of files and lastly it backs up permission.

This again has problems since all data are encrypted for encrypted user. Obb files aren't a problem because Obb files only change when you update an app.

> I'll send them via DM if you're interested.

Sure!

**Comment by Cyberavater** [comment]
> Sure!

Sent.

>The only unencrypted file is the metadata file.

Why not add size info in this metadata file for size check? since size check is a good way to determine the difference in most cases (even checksums if needed?).

>Version checks should be covered by 1(c).

Ya, I know, but it's still important for 1(d), as it's a superset of 1(c).

**Comment by MuntashirAkon** [comment]
> Why not add size info in this metadata file for size check? since size check is a good way to determine the difference in most cases (even checksums if needed?).

Based on the documents you've sent me, it is clear that they're checking app sizes. I think backup time could also utilised in this manner (to find the last launch date for an app). This will need usage access permission.

It would be better, if you have sent me some metadata files (if stored). They would also help a lot.

**Comment by Cyberavater** [comment]
If checking app sizes is enough then I guess no need to introduce 'last launch time' just yet, first only implement time check in debug build and see how it goes.

**Comment by MuntashirAkon** [comment]
I'm finalising the features which can be described as follows:
1. **Backup apps:** 1-click option regarding the backup of apps, it has the following options in the submenu:
  a. Backup all apps
  b. Backup apps without any previous backups
  c. Backup apps with changes: update date, size, last launch data (AM aims to minimise publicly available metadata which means it cannot store sensitive metadata in plain text)
2. **Restore apps:** 1-click option regarding the restoration of already backed up apps, it has the following options in the submenu:
  a. Restore all apps
  b. Restore apps that are not currently installed
  c. Restore apps whose backup version is higher than the installed version

This needs one extra metadata, namely `size`.

**Comment by Cyberavater** [comment]
Add 'Sort' and 'Filter' options as available in onboard;

After selecting 1(a) if the user selects the "User apps" filter, It'd only backup user apps leaving the system apps.

As I see, the "Apps with backup" filter here would have the potential to remove 1(c). As for 1(d) on previous comment, it should be already added in 1(c).

Here, the filter option would also contain custom labels, #236

**Comment by Cyberavater** [comment]
Additionally, It'd be nice to have the apps' icon here too, as it's easier to detect apps by an icon rather than reading the name.

**Comment by MuntashirAkon** [comment]
> Additionally, It'd be nice to have the apps' icon here too, as it's easier to detect apps by an icon rather than reading the name.

Adding app icons there is a bit problematic but I'll see if I can do it.

**Comment by MuntashirAkon** [comment]
> the "Apps with backup" filter here would have the potential to remove 1(c).

You mean 1(e)?

**Comment by Cyberavater** [comment]
Oh, As I see it, 1(e) (Redo all existing backups (opposite of b)) is pretty much useless, i.e should not be implemented.
As a general rule, not only for 1(c), before backing up an app, AM should already check if it's already backed up, if yes, then then it should perform the necessary checks (mentioned for 1(c) including 1(d). here, https://github.com/MuntashirAkon/AppManager/issues/136#issuecomment-745031032) and selectively redo backup only for the changed component, i.e,
If I single-ly backup Mobile Legends from Onboard (Not 1-click). AM would still check if there already exists a backup, if yes, it'd perform the necessary checks per components, and finding only change in the 'Data' component, it'd skip the 4GB+ "External Data" component and only update the corresponding component, this would greatly decrease the time while doing repeated backups.

**Comment by Cyberavater** [comment]
> As a general rule, not only for 1(c), before backing up an app, AM should already check if it's already backed up, if yes, then then it should perform the necessary checks (mentioned for 1(c) including 1(d). here, #136 (comment)) and selectively redo backup only for the changed component, i.e,
If I single-ly backup Mobile Legends from Onboard (Not 1-click). AM would still check if there already exists a backup, if yes, it'd perform the necessary checks per components, and finding only change in the 'Data' component, it'd skip the 4GB+ "External Data" component and only redo the corresponding component and update the backup, this would greatly decrease the time while doing repeated backups.

This method followed by Swift Backup.

I have another idea, that is the Inverse filter!  With this, the user can make all the rule himself and do all things in the Onboard page that which looks more elegant. (more detailed; color icon etc)
i.e.
If the user wants to emulate 1(b), he would select;
Backed up apps in the inverse filter,

And with the addition of some more filters (ie. "Installed", etc) , I think it is possible to emulate every case available there. (If 1(c) is implemented as general, as mentioned above).

**Comment by MuntashirAkon** [comment]
> I have another idea, that is the Inverse filter! With this, the user can make all the rule himself and do all things in the Onboard page that which looks more elegant. (more detailed; color icon etc)
> i.e.
> If the user wants to emulate 1(b), he would select;
> Backed up apps in the inverse filter,
> 
> And with the addition of some more filters (ie. "Installed", etc) , I think it is possible to emulate every case available there. (If 1(c) is implemented as general, as mentioned above).

This isn't possible. I don't want to clutter the filtering option with all sorts of backup specific stuffs that can be performed with a profile or via 1-click option. Keep in mind that, backup/restore is just _one_ of the features that AM offers, it took only 2200 lines of Java code to implement it (if you compare it to other features: app ops and profiles have 2000+ lines each).

> If I single-ly backup Mobile Legends from Onboard (Not 1-click). AM would still check if there already exists a backup, if yes, it'd perform the necessary checks per components, and finding only change in the 'Data' component, it'd skip the 4GB+ "External Data" component and only redo the corresponding component and update the backup, this would greatly decrease the time while doing repeated backups.

So, if I understand correctly, explicit checks on source and data is necessary here. In order to do that, AM has to store the hash of the data during the backup process in the database. When AM checks for a change, it will make hash of the current data and match them against the one stored in the database. This would be the most efficient and secured option to do this. Metadata is not the place to do it since it metadata are intended to be portable and universal.

Now, to finalise 1(c), the following will be checked for each app (sorted in terms of priority):
1. App version code
2. Last update data (against last backup date)
3. Last launch date (against last backup date)
4. Integrity check as in 1(d)
5. File hash checks

Of the above list, (5) will be done in all cases to determine backup options. User can still set backup options but they will be removed based on the checks.

**Comment by MuntashirAkon** [comment]
We won't be determine backup options for (5) for now. This will be supplied as a patch in v2.6.1.

**Comment by MuntashirAkon** [comment]
Depends on #397

**Comment by MuntashirAkon** [comment]
Instead of file hash checks, timestamp based modification detection is much more effective and faster, thus, implemented instead (7b562d8742ae053534665cdd34d55831f64e796c).
